### DISCUSSION TITLE
Remote control of "failures"

### DISCUSSION THREAD

**Discussion by MikeMcPherson** [alanbjohnston/CubeSatSim#238] [Ideas]
Hi all

As part of incorporating the CubeSatSim into the Aerospace Engineering Spacecraft Design capstone course at UVA, I am looking at adding the ability to "fail" components and systems remotely.  This capability would be used to teach spacecraft characterization, fault analysis, and fault mitigation.  Using the remote interface, the instructor or a fellow student would simulate failures.  The students, using the baseline performance data collected during the testing and characterization phase, would be tasked with creating a testing plan, identifying the failures, and proposing a mitigation plan.

I'm thinking a Web interface for commanding failures.  That would allow relatively untrained operators (like fellow students) take the command role, and would make monitoring data available to the command operator.  This would, of course, require some sort of two-way wireless connectivity with the CubeSatSim.  A Pi Zero W or Pico W used as the main MCU would do the trick.  Alternatively, a LoRa radio could be added to the payload board.  Those are just the first possibilities that pop into my head.

Possible failure modes could include:

- Solar panels (fail open, shorted, reduced output)
- Batteries (fail open, shorted, reduced output, reduced capacity failure to charge)
- Voltage/current sensors (fail completely, produce erroneous results)
- "Radio" (fail completely, reduced signal strength, various levels of noise or intentional interference, stuck bit(s))
- Main MCU (fail completely, intermittent failure, computation errors, excessive current draw, memory failure) 
- Payload board (failure of sensors, failure of MCU, excessive current draw)

For most of the failure modes, the ability to command an intermittent, progressive, or cascading failure would be useful.

I'm looking for brainstorming from others who might be interested in this sort of capability.  If you are interested, or if you have other thoughts about failure modes that would be interesting, please chime in.

Thanks.  Mike KQ9P

**Comment by msamis1** [comment]
Hi Mike,
I am on the Educational Materials team that Alan mentioned in his post.  We have been working on lesson plans/experiments for the CubSatSim Thank you for the great suggestions and your work. I believe that creating simulated failures in software is a great idea! All of the failure modes that you mention we have experienced in reality in some form or other. Using a web-based interface to remotely invoke these failures is also a great idea! Intermittent and cascading failures seem to occur quite a bit when building and testing the CubeSatSim.  However, I have found that often the issue is actually not with the CubeSatSim but with the ground station configuration. I am not suggesting creating a failure option for the ground station. I think that option for CubeSatSim is the best, especially for engineering students. 
Thanks again for all your work and the suggestions.
Mark Samis, KD2XS

**Comment by alanbjohnston** [comment]
I like this as well, as an extra optional capability.  I'd like to hear more about it.

**Comment by MikeMcPherson** [comment]
Thanks for the support, folks.  I plan on doing this, but it won't be right away.  Too many commitments...  When I have time to get back to this, I'll do some interface mockups and post them here.

Mike KQ9P

**Comment by W7DGP** [comment]
Having taught troubleshooting in the past, I believe this is a great idea. Looking at possible solutions, I can see that a web interface to LoRa radio, some PHP would be a ton of work. In that most of your faults seem to be a simple connection I would suggest a small relay board and a four-button fob might get you started fairly quickly. Nothing like troubleshooting gets that critical thinking engine running!

Larry W7DGP
